modules/gdnative/nativearvr/arvr_interface_gdnative.cpp

```diff
@@ -0,0 +1,286 @@
+/*                      https://godotengine.org                          */
+/*************************************************************************/
+/* Copyright (c) 2007-2017 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2017 Godot Engine contributors (cf. AUTHORS.md)    */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "arvr_interface_gdnative.h"
+#include "servers/visual/visual_server_global.h"
+
+ARVRInterfaceGDNative::ARVRInterfaceGDNative() {
+	// testing
+	printf("Construct gdnative interface\n");
+
+	// we won't have our data pointer until our library gets set
+	data = NULL;
+}
+
+ARVRInterfaceGDNative::~ARVRInterfaceGDNative() {
+	printf("Destruct gdnative interface\n");
+
+	if (is_initialized()) {
+		uninitialize();
+	};
+
+	// cleanup after ourselves
+	cleanup();
+}
+
+void ARVRInterfaceGDNative::cleanup() {
+	if (data != NULL) {
+		void *parameters[1];
+		parameters[0] = (void *)this;
+		const_cast<GDNative *>(library.ptr())->call_native_raw("arvr_call_destructor", "godot_arvr_destructor", data, 1, parameters, NULL);
+
+		data = NULL;
+	};
+
+	if (library.is_valid()) {
+		library->terminate();
+	};
+};
+
+void ARVRInterfaceGDNative::set_gdnative_library(Ref<GDNativeLibrary> p_library) {
+	if (library.is_null()) {
+		library.instance();
+	} else {
+		cleanup();
+	};
+
+	library->set_library(p_library);
+	library->initialize();
+
+	// Now we do our constructing...
+	void *parameters[1];
+	parameters[0] = (void *)this;
```

Comment:
I actually think it should be enough to just pass the `this` pointer on the constructor and nowhere else, if the library needs it it can save it in the `p_data`.

I did that with the nativescript functions too, but as it turns out it's not really used in any bindings, they all just save it in the custom data and then never touch those parameters again. You can keep it, just saying that it probably won't end up being as useful as it seems.

But maybe it turns out to be a great help, who knows, just mentioning it here :)

Reply:
I figured these may have been useful if you need to do calls on any of the other member functions of the object, actually one thing that is still missing is updating the tracking status which is an instance variable of the superclass definitely process will end up needing it. But yeah for many others its more there for consistency then it being useful

Reply:
I'm changing this as well, if they want the pointer they can always copy it into the data structure (I'll add that into the example implementation)